src/checkpoint.hh

```diff
@@ -0,0 +1,1246 @@
+       }
+     }
+
+     if (supported && (strings || subseq || user_def)) {
+       /*
+          generate accessors for String/Subsequence/user-defined types;
+          these are needed to access these objects efficiently during
+          (de)serialization of the checkpointed archives;
+          for Strings: only works if String objects are wrapped in a List_Ref
+          object, so return false if List_Ref is not part of the type
+       */
+       if (strings && !list_ref) {
+         supported = false;
+       } else {
+         std::vector<std::string> type_accessor;
+         gen_type_accessors(table_type, type_accessor);
+       }
+     }
+
+     return supported;
+  }
+
+  void macros(Printer::Base &stream) {
+     stream << "#define CHECKPOINTING_INTEGRATED" << endl << endl;
+     if (list_ref) {
+       // set macro if tables contain List_Ref type
+       stream << "#define LIST_REF" << endl << endl;
+     }
+     if (strings) {
+       // define macros for easy/efficient access to String/Block objects
+       for (size_t i = 0; i < string_type_accessors.size(); i++) {
+         if (!(string_type_accessors[i].empty())) {
+           stream << "#define S" << i+1 << " "
+                  << string_type_accessors[i]
+                     .substr(1, string_type_accessors[i].npos)<< endl;
+           stream << "#define B" << i+1 << " "
+                  << block_type_accessors[i]
+                     .substr(1, block_type_accessors[i].npos)<< endl;
+         }
+       }
+       stream << endl;
+     }
+     if (subseq) {
+       // define macros for easy/efficient access to Subsequence objects
+       for (size_t i = 0; i < subseq_type_accessors.size(); i++) {
+         if (!(subseq_type_accessors[i].empty())) {
+           stream << "#define SUBSEQ" << i+1 << " "
+                  << subseq_type_accessors[i]
+                     .substr(1, subseq_type_accessors[i].npos)<< endl;
+         }
+       }
+       stream << endl;
+     }
+  }
+
+  void include(Printer::Base &stream, const nt_tables &tables) {
+     stream << "extern \"C\" {" << endl;
+     stream << indent() << "#include <unistd.h>" << endl;
+     stream << indent() << "#include <sys/resource.h>" << endl;
+     stream << "}" << endl;
+     stream << "#include \"boost/serialization/vector.hpp\"" << endl;
+     stream << "#include \"boost/serialization/utility.hpp\"" << endl;
+     stream << "#include \"boost/serialization/access.hpp\"" << endl;
+     stream << "#include \"boost/archive/binary_iarchive.hpp\"" << endl;
+     stream << "#include \"boost/archive/binary_oarchive.hpp\"" << endl;
+     stream << "#include \"boost/filesystem.hpp\"" << endl;
+     stream << "#include <atomic>" << endl;
+     stream << "#include <ctime>" << endl;
+     stream << "#include <unordered_map>" << endl;
+     stream << "#include <mutex>" << endl;
+     stream << "#include <thread>" << endl << endl;
+  }
+
+  void restore_string_links(Printer::Base &stream, const nt_tables &tables) {
```

Comment:
this was a beast. It might be worth to leave a comment WHY strings a such a problem and explain Georg's logic for storing strings this way

Reply:
It sure was. I left a comment that roughly explains the problem and how its solved.